diff --git a/zapzap/__init__.py b/zapzap/__init__.py
index 85707b0..7a0b117 100644
--- a/zapzap/__init__.py
+++ b/zapzap/__init__.py
@@ -44,7 +44,10 @@ path_tmp = os.path.join(
         QStandardPaths.StandardLocation.AppLocalDataLocation), __appname__, 'tmp'
 )
 if not os.path.exists(path_tmp):
-    os.makedirs(path_tmp)
+    try:
+        os.makedirs(path_tmp)
+    except:
+        print("Not Writable")
 
 # Path translations
 po_path = os.path.join(abs_path, 'po')
@@ -59,7 +62,10 @@ DATABASE_DIR = os.path.join(
 )
 
 if not os.path.exists(DATABASE_DIR):
-    os.makedirs(DATABASE_DIR)
+    try:
+        os.makedirs(DATABASE_DIR)
+    except:
+        print("Not Writable")
 
 DATABASE_FILE = os.path.join(DATABASE_DIR, 'zapzap.db')
 
