diff --git a/modules/gtk/hm.nix b/modules/gtk/hm.nix
index 24e2c66..e337cf0 100644
--- a/modules/gtk/hm.nix
+++ b/modules/gtk/hm.nix
@@ -60,7 +60,7 @@ mkTarget {
             package = pkgs.adw-gtk3;
             name = "adw-gtk3";
           };
-          gtk.gtk4.theme = config.gtk.theme;
+          gtk.gtk4.theme = lib.mkForce config.gtk.theme;
 
           xdg.configFile = {
             "gtk-3.0/gtk.css".source = finalCss;
